Userland/Services/SpiceAgent/SpiceAgent.cpp

```diff
@@ -40,6 +41,59 @@ ErrorOr<void> SpiceAgent::start()
     auto capabilities_message = AnnounceCapabilitiesMessage::create(m_capabilities);
     TRY(this->send_message(capabilities_message));
 
+    GUI::Clipboard::the().on_change = [this](auto const& mime_type) {
+        auto result = this->on_clipboard_update(String::from_deprecated_string(mime_type).release_value_but_fixme_should_propagate_errors());
+        if (result.is_error()) {
+            dbgln("Failed to inform the spice server of a clipboard update: {}", result.release_error());
+        }
+    };
+
+    return {};
+}
+
+ErrorOr<void> SpiceAgent::on_clipboard_update(String const& mime_type)
+{
+    // If we just copied something to the clipboard, we shouldn't do anything here
+    if (m_just_updated_clipboard) {
+        m_just_updated_clipboard = false;
+        return {};
+    }
+
+    // If the clipboard has just been cleared, we shouldn't send anything
+    if (mime_type.is_empty()) {
+        return {};
+    }
+
+    // Notify the spice server about new content being available
+    auto clipboard_data_type = TRY(from_mime_type(mime_type));
+    auto message = ClipboardGrabMessage::create({ clipboard_data_type });
+    TRY(this->send_message(message));
+
+    return {};
+}
+
+ErrorOr<void> SpiceAgent::send_clipboard_contents(ClipboardDataType data_type)
+{
+    auto data_and_type = GUI::Clipboard::the().fetch_data_and_type();
+    auto requested_mime_type = TRY(to_mime_type(data_type));
+
+    // We have an exception for `image/x-serenityos`, where we treat it as a PNG when talking to the spice server
+    auto is_serenity_image = data_and_type.mime_type == "image/x-serenityos" && data_type == ClipboardDataType::PNG;
+    if (!is_serenity_image && requested_mime_type.to_deprecated_string() != data_and_type.mime_type) {
+        // If the requested mime type doesn't match what's on the clipboard, we won't send anything back
+        return Error::from_string_literal("Requested mime type doesn't match the clipboard's contents!");
+    }
+
+    // If the mime type is `image/x-serenityos`, we need to encode the image that's on the clipboard as a PNG
```

Comment:
If we have raw bitmaps in x-serenityos (if that's what it is?) is there a more ... raw? image format that the spice protocol supports we could use so that we have to do less work here? It would be nice to not have to decode or encode images in this process that's responsible for clipboard communication. We split that out into the ImageDecoder service in the Browser for security reasons (image decode/encode is notoriously hard to make fuzz-resistant).

Reply:
Unfortunately, the protocol only  supports PNG, JPEG, BMP and TIFF.